core/src/plugins/kubernetes/results-cache-fs.ts

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (C) 2018-2024 Garden Technologies, Inc. <info@garden.io>
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+import type { CacheStorage, SchemaVersion } from "./results-cache-base.js"
+import { CacheStorageError } from "./results-cache-base.js"
+import fsExtra, { pathExists } from "fs-extra"
+import { join } from "path"
+import writeFileAtomic from "write-file-atomic"
+import { CACHE_DIR_NAME } from "../../constants.js"
+import type { Log } from "../../logger/log-entry.js"
+import { isErrnoException } from "../../exceptions.js"
+import { RootLogger } from "../../logger/logger.js"
+import type { JsonObject } from "type-fest"
+import { listDirectory } from "../../util/fs.js"
+import moment from "moment/moment.js"
+import { lstat } from "fs/promises"
+
+const { ensureDir, readFile, remove } = fsExtra
+
+class LocalFileSystemCacheError extends CacheStorageError {
+  override type = "local-fs-cache-storage"
+}
+
+export const FILESYSTEM_CACHE_EXPIRY_DAYS = 7
+
+/**
+ * Very simple implementation of file-system based cache
+ * to be used as a fallback storage for kubernetes Run and Test results.
+ *
+ * It uses cache keys to name and create JSON files,
+ * and stores the values in the JSON files in a configurable cache directory.
+ *
+ * This class holds the minimal in-memory state,
+ * because each Run and Test result
+ * is usually read and written only once per Garden command execution.
+ */
+export class SimpleLocalFileSystemCacheStorage implements CacheStorage {
+  private readonly cacheDir: string
+  private readonly schemaVersion: SchemaVersion
+  private readonly cacheExpiryDays: number
+  private readonly log: Log
+
+  private constructor({
+    cacheDir,
+    schemaVersion,
+    cacheExpiryDays,
+  }: {
+    cacheDir: string
+    schemaVersion: SchemaVersion
+    cacheExpiryDays: number
+  }) {
+    this.cacheDir = cacheDir
+    this.schemaVersion = schemaVersion
+    this.cacheExpiryDays = cacheExpiryDays
+    this.log = RootLogger.getInstance().createLog({ name: "fs-cache" })
+  }
+
+  public static async getInstance(params: {
+    cacheDir: string
+    schemaVersion: SchemaVersion
+    cacheExpiryDays: number
+  }): Promise<SimpleLocalFileSystemCacheStorage> {
+    const instance = new SimpleLocalFileSystemCacheStorage(params)
+    await instance.invalidate()
```

Comment:
Do we need an async lock to prevent invalidating the same directory multiple times at once?
Do we really want to wait until invalidation completes before moving on? I think it would be absolutely acceptable in this instance to not await the promise.
```suggestion
    void instance.invalidate()
```